Add support for Pickled Graphite over UDP

michaelsembwever opened the pull request:
Redo of https://github.com/dropwizard/metrics/pull/714

> Pulled the logic to create the Pickled data format into its own class (GraphitePickler).
> Adjusted PickledGraphite to use the GraphitePickler class.
> Created PickledGraphiteUDP class which is to PickledGraphite what GraphiteUDP is to Graphite.

arteam commented on metrics-graphite/src/main/java/com/codahale/metrics/graphite/PickledGraphiteUDP.java:
```diff
@@ -0,0 +1,203 @@
+import java.util.LinkedList;
+import java.util.List;
+
+public final class PickledGraphiteUDP implements GraphiteSender {
```
Could you please add a unit test for `PickledGraphiteUDP`?

arteam commented on metrics-graphite/src/main/java/com/codahale/metrics/graphite/PickledGraphiteUDP.java:
```diff
@@ -0,0 +1,203 @@
+                ByteBuffer byteBuffer = ByteBuffer.allocate(4 + payload.length).putInt(payload.length).put(payload);
+
+                datagramChannel.send(byteBuffer, address);
+                this.failures = 0;
```
Could you please elaborate why we need to reset the counter? In `PickledGraphite` we don't do that.

arteam commented on metrics-graphite/src/main/java/com/codahale/metrics/graphite/PickledGraphiteUDP.java:
```diff
@@ -0,0 +1,203 @@
+    @Override
+    public void close() throws IOException {
+        datagramChannel.socket().close();
+        datagramChannel = null;
```
I think it's better to dereference the channel in a finally block in case the socket failed to close.

michaelsembwever commented:
The code commits are from @philwill-nap, someone i do not know, so I'd like to hold them intact to represent the authorship correctly.

I will add an additional commit addressing the comments and referencing this PR.

arteam commented:
> so I'd like to hold them intact to represent the authorship correctly.

Sure, you can push your changes as separate commits.

arteam commented:
So, we have only pme review point to address. If you contribute a unit test for `PickledGraphiteUDP`, I will happily merge the pull request.

michaelsembwever commented:
>  If you contribute a unit test for PickledGraphiteUDP …

will do, but will need to come back to it soon, being holiday season n all :-)

arteam commented:
No rush, thanks for taking your time to contribute!

iksaif commented:
FYI: pickle is usually less performant that lines (uses more CPU) and will produce bigger datagrams that are more likely to be silently dropped (because UDP)
